src/Scriban/Syntax/Expressions/ScriptIndexerExpression.cs

```diff
@@ -130,17 +130,22 @@ private object GetOrSetValue(TemplateContext context, object valueToSet, bool se
             {
                 var accessor = context.GetMemberAccessor(targetObject);
                 var indexAsString = context.ObjectToString(index);
+                var itemAccessor = accessor as IItemAccessor;
 
                 if (setter)
                 {
-                    if (!accessor.TrySetValue(context, Index.Span, targetObject, indexAsString, valueToSet))
+                    var itemSuccess = itemAccessor?.ItemType == index?.GetType() && itemAccessor.TrySetItem(context, Index.Span, targetObject, index, valueToSet);
```

Comment:
Should probably use `context.ToObject` to `itemAccessor?.ItemType` instead to convert the index only if ItemType is not null and use the item accessor only in this case.

Reply:
(a source indexer value could be e.g long while the indexer is int, automatic casting is usually done in scriban for such case)

Reply:
Updated